### PR TITLE
Cary - L251.en_ssp_nonco2

### PR DESCRIPTION
**Pull request by CLynchy**
#267

### CODE REVIEW

File: R/constants.R [hunk 1]
```diff
@@ -106,9 +106,12 @@ CONV_KG_TO_TG <- 1e-9
 CONV_M3_BM3 <- 1e-09
 CONV_MILLION_M3_KM3 <- 1e-03
 
+
+
 CONV_M2_ACR <- 0.0002471058
```
Comment (cahartin) [inline R/constants.R]: comment, please.

Comment (cahartin) [inline R/constants.R]: comment, please.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 2]
```diff
@@ -1,27 +1,32 @@
 #' module_emissions_L251.en_ssp_nonco2
 #'
-#' Briefly describe what this chunk does.
+#' This chunk produces regional non-CO2 emissions coefficient data for SSPs 1/5, 2, and 3/4 as well as a GDP control.
```
Comment (cahartin) [inline R/zchunk_L251.en_ssp_nonco2.R]: can you change these to 1.5 and 3.4

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 3]
```diff
@@ -1,27 +1,32 @@
...
 #'
 #' @param command API command to execute
 #' @param ... other optional parameters, depending on command
 #' @return Depends on \code{command}: either a vector of required inputs,
 #' a vector of output names, or (if \code{command} is "MAKE") all
 #' the generated outputs: \code{L251.ctrl.delete}, \code{L251.ssp15_ef}, \code{L251.ssp2_ef}, \code{L251.ssp34_ef}, \code{L251.ssp15_ef_vin}, \code{L251.ssp2_ef_vin}, \code{L251.ssp34_ef_vin}. The corresponding file in the
 #' original data system was \code{L251.en_ssp_nonco2.R} (emissions level2).
-#' @details Describe in detail what this chunk does.
+#' @details This section takes in the non-CO2 emissions factors for SSP 1.5, 2, and 3/4 across sectors.
```
Comment (cahartin) [inline R/zchunk_L251.en_ssp_nonco2.R]: 3.4?

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 4]
```diff
@@ -35,106 +40,232 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
-      add_units("units") %>%
-      add_comments("comments describing how data generated") %>%
-      add_comments("can be multiple lines") %>%
+# This section takes in the non-CO2 emissions factors for SSP 1/5, 2, and 3/4 across sectors and
+# interpolates across years 2010-2100 in 5 year segments.
+
+    L161.SSP15_EF %>%
+      # Input data only contains 3 future years.  Need to include years between 2010-2100 in 5 year segments,
+      # and interpolate between each time segment.
+      tidyr::complete(year = SSP_FUTURE_YEARS,
```
Comment (cahartin) [inline R/zchunk_L251.en_ssp_nonco2.R]: @bpbond : do we need the `tidyr::`?

File: inst/extdata/temp-data-inject/L201.nonghg_steepness.csv [hunk 5]
```diff
@@ -1,8 +1,8 @@
-INPUT_TABLE
```
Comment (bpbond) [inline inst/extdata/temp-data-inject/L201.nonghg_steepness.csv]: What's going on with this file? It can't be read in by the new data system.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 6]
```diff
@@ -35,106 +40,232 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
+      left_join_error_no_match(A_regions, by = c("GCAM_region_ID")) %>%
+      # Emission coefficients values are too long, so we round to the 10th decimal point.
+      mutate(emiss.coeff = round(value, 10)) %>%
+      ungroup() %>%
+      # Discard columns that are not needed.
+      select(-MAC_region, -bio_N2O_coef, -SO2_name, -GAINS_region, -GCAM_region_ID, -agg_sector, -value) ->
+    L251.ssp15_ef
+
+    # Reorder data.
+    L251.ssp15_ef <- L251.ssp15_ef[c(6,3:5,1:2,7)]
```
Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: Ugh, why?!? Very fragile. Not necessary surely.

File: R/constants.R (hunk 1 above)
Comment (CLynchy) [inline R/constants.R]: Sorry, I forgot to check constants.  I don't know why there are so many blank lines.  Everytime I update from master something in the "constants.R" looks crazy.

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 6 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: I kept getting errors, and I thought "it can't hurt".
Should I remove?

File: inst/extdata/temp-data-inject/L201.nonghg_steepness.csv (hunk 5 above)
Comment (CLynchy) [inline inst/extdata/temp-data-inject/L201.nonghg_steepness.csv]: I will copy the file from the old system into this folder.  I may have opened this file in Excel (accidently!) which does crazy stuff to the file.

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 2 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: I will fix all the 1/5 = 1.5 and 3/4 = 3.4

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 6 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: Reordering has been removed :*(

**Comment by bpbond** [comment]
@CLynchy note your commit 46949ec broke something

**Comment by CLynchy** [comment]
Had to adjust that **** L201* file again.  When I copied it over to the new folder I didn't comment out the first 3 lines, which was needed.

**Comment by bpbond** [comment]
@ssmithClimate Corinne is traveling/out of town, and I'm not sure when she'll be able to review this. Steve, would you have time and interest to help? Just need someone to go through Cary's code and evaluate it for clarity and algorithm, disciplinary-wise, flagging things as necessary for follow-up. Thanks.

**Comment by ssmithClimate** [comment]
@bpbond Sure, will do.

File: inst/extdata/emissions/A_regions.csv [hunk 7]
```diff
@@ -0,0 +1 @@
...
23,Russia,Russia,0.004930186,SO2_4,Russia+
24,South Africa,Africa,0.009445905,SO2_3,South Africa
25,South America_Northern,Central and South America,0.008163204,SO2_3,Rest South America
26,South America_Southern,Central and South America,0.008163204,SO2_3,Rest South America
27,South Asia,Asia,0.0095297,SO2_2,Rest South Asia
28,South Korea,Asia,0.007774893,SO2_2,Korea
29,Southeast Asia,Asia,0.0095297,SO2_2,Southeastern Asia
30,Taiwan,China,0.0095297,SO2_2,China+
31,Argentina,Central and South America,0.008163204,SO2_3,Rest South America
32,Colombia,Central and South America,0.008163204,SO2_3,Rest South America
```
Comment (rplzzz) [inline inst/extdata/emissions/A_regions.csv]: Line endings on this file are messed up.  Please fix them.

File: inst/extdata/temp-data-inject/L201.nonghg_steepness.csv [hunk 8]
```diff
@@ -1,7 +1,7 @@
...
-Variable ID
-GDPCtrlSteep
-
+# INPUT_TABLE
+# Variable ID
+# GDPCtrlSteep
+#
```
Comment (rplzzz) [inline inst/extdata/temp-data-inject/L201.nonghg_steepness.csv]: What do these header lines mean?  They're kind of cryptic.

File: tests/testthat/comparison_data/emissions/L251.ctrl.delete.csv [hunk 9]
```diff
@@ -0,0 +1,5516 @@
+INPUT_TABLE
+Variable ID
+DelEmCtrl
+
```
Comment (rplzzz) [inline tests/testthat/comparison_data/emissions/L251.ctrl.delete.csv]: These header lines are also cryptic, and unlike in `L201.nonghg_steepness.csv` they're not prefixed with `#`.  Are these supposed to be comments, or do they mean something to the data system and get processed automatically?  If the latter, then do the comment characters need to be removed from `L201...` above?

Comment (rplzzz) [inline tests/testthat/comparison_data/emissions/L251.ctrl.delete.csv]: Also, there are a couple more csv files in the changeset that are too large for github to render the diffs on them.  They need to be checked to see that they have correct line endings and for correctness of these metadata lines.

Also, doesn't the data system require a standard set of metadata comments for input files?

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 4 above)
Comment (rplzzz) [inline R/zchunk_L251.en_ssp_nonco2.R]: I usually include it unless it's something I'm using all the time.  It never hurts to let someone reading the code know where a function comes from.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 10]
```diff
@@ -35,106 +40,214 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
+      group_by(GCAM_region_ID, Non.CO2, supplysector, subsector, stub.technology, agg_sector) %>%
+      mutate(value = approx_fun(year, value))  %>%
+      # Now add regional information.
+      left_join_error_no_match(A_regions, by = c("GCAM_region_ID")) %>%
+      # Emission coefficients values are too long, so we round to the 10th decimal point.
+      mutate(emiss.coeff = round(value, 10)) %>%
+      ungroup() %>%
+      # Discard columns that are not needed.
+      select(-MAC_region, -bio_N2O_coef, -SO2_name, -GAINS_region, -GCAM_region_ID, -agg_sector, -value) ->
+    L251.ssp15_ef
```
Comment (rplzzz) [inline R/zchunk_L251.en_ssp_nonco2.R]: I would include a note here indicating what columns are in the resulting output (the `select` above tells us what's _not_ in, but that's not so useful).  If all of these tables are going to have the same set of columns, then you could just expand the comment at lines 64-65 to include this information and note that it applies to all of these tables.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 11]
```diff
@@ -35,106 +40,214 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
+      select(-MAC_region, -bio_N2O_coef, -SO2_name, -GAINS_region, -GCAM_region_ID, -agg_sector, -value) ->
+    L251.ssp34_ef
+
+# Create a control table where year 2010 is the control period and renamed 1975, discarding all other years.
+     L251.ssp2_ef %>%
+      select(-emiss.coeff) %>%
+      subset(year == min(year)) %>%
+      mutate(year = 1975,
+             ctrl.name ="GDP_control") ->
+      L251.ctrl.delete
```
Comment (rplzzz) [inline R/zchunk_L251.en_ssp_nonco2.R]: No "alt"?

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 12]
```diff
@@ -35,106 +40,214 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
+       # Future emission coefficient name is a descriptor, and is constant.
+       mutate(future.emiss.coeff.year = year,
+              future.emiss.coeff.name = "SSP_GAINS",
+              # Previous year column that only includes the electricity supply sector is now a constant.
+              year = 1975)->
+      L251.ssp15_ef_vin
+
+     L251.ssp2_ef %>%
+       # Isolate the "electricity" supply sector.
+       subset(supplysector == "electricity") %>%
```
Comment (rplzzz) [inline R/zchunk_L251.en_ssp_nonco2.R]: You should use `filter` for this operation and other similar operations throughout.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 13]
```diff
@@ -35,106 +40,214 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
+     L251.ctrl.delete <- rename_SO2( L251.ctrl.delete, A_regions, FALSE )
+     L251.ssp15_ef <- rename_SO2( L251.ssp15_ef, A_regions, FALSE )
+     L251.ssp2_ef <- rename_SO2( L251.ssp2_ef, A_regions, FALSE )
+     L251.ssp34_ef <- rename_SO2( L251.ssp34_ef, A_regions, FALSE )
+     L251.ssp15_ef_vin <- rename_SO2( L251.ssp15_ef_vin, A_regions, FALSE )
+     L251.ssp2_ef_vin <- rename_SO2( L251.ssp2_ef_vin, A_regions, FALSE )
+     L251.ssp34_ef_vin <- rename_SO2( L251.ssp34_ef_vin, A_regions, FALSE )
+
+# This section deletes GDP control functions that exist for the supply sector in all
+# regions for non-CO2 emission species for the 1975 control period.
```
Comment (rplzzz) [inline R/zchunk_L251.en_ssp_nonco2.R]: This comment is really hard to follow (the prepositional phrases appear to be nested four deep).  Could you reword it for clarity?  Also, a note as to _why_ we need to do this would be helpful.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 14]
```diff
@@ -35,106 +40,214 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
+                 by = c("region", "supplysector", "subsector", "stub.technology", "Non.CO2")) ->
+       L251.ctrl.delete
+
+# ===================================================
+# Produce outputs
+     # No flags are neccessary because old data is in 'long' format.
+
+    L251.ctrl.delete %>%
+      add_title("GDP control of regional non-CO2 emissions.") %>%
+      add_units("NA") %>%
```
Comment (rplzzz) [inline R/zchunk_L251.en_ssp_nonco2.R]: Let's not use the string "NA".  It's probably harmless here, but it's best to avoid it anyhow.  I've been using "unitless" for pure numbers.

File: tests/testthat/comparison_data/emissions/L251.ctrl.delete.csv (hunk 9 above)
Comment (bpbond) [inline tests/testthat/comparison_data/emissions/L251.ctrl.delete.csv]: I think this file must have been included by mistake?

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 14 above)
Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: Good point.

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 13 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: I roll deep, what can I say?  I will tweak this sentence, but I can honestly say I don't know why this step is done.  The original code leaves much to the imagination.

Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: >I don't know why this step is done

Then check with Page please!

Comment (ssmithClimate) [inline R/zchunk_L251.en_ssp_nonco2.R]: Kate probably did this one, but I can help with this.

File: inst/extdata/temp-data-inject/L201.nonghg_steepness.csv (hunk 8 above)
Comment (CLynchy) [inline inst/extdata/temp-data-inject/L201.nonghg_steepness.csv]: I assume that this file will be produced by another persons chunk.  It is an input file for me, and has been a pain to work with.  I intensely dislike the L201* file.  All that being said, should I make edits to this input file if it is going to be rewritten by someone else?

Comment (bpbond) [inline inst/extdata/temp-data-inject/L201.nonghg_steepness.csv]: I did not know that a level 2 file could feed back into Level 1--this is just weird.

>should I make edits to this input file 

No. I'm impressed that `get_data` was even able to read it, honestly. Let me check out your branch and take a look at this later.

Comment (CLynchy) [inline inst/extdata/temp-data-inject/L201.nonghg_steepness.csv]: The chunk this pull request is about is a Level 2, right?

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 11 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: ?

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 14 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: There are no numbers in this output file.  Everything is in string/character, with the exception of the "year" column (which is a constant).

File: inst/extdata/temp-data-inject/L201.nonghg_steepness.csv (hunk 8 above)
Comment (bpbond) [inline inst/extdata/temp-data-inject/L201.nonghg_steepness.csv]: Ah, good point.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 15]
```diff
@@ -1,27 +1,32 @@
...
+#' First, create data that spans the years 2010-2100 in five year increments by interpolation of input data.
+#' Next, add emissions controls for future years of vintaged technologies for SSP emission factors.
+#' Then, add columns that have regional SO2 emission species.
+#' A GDP control of regional non-CO2 emissions in all regions is also created.
 #' @importFrom assertthat assert_that
 #' @importFrom dplyr filter mutate select
 #' @importFrom tidyr gather spread
-#' @author YourInitials CurrentMonthName 2017
+#' @author CDL May 2017
 #' @export
```
Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: Remove `@export` line

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 16]
```diff
@@ -35,106 +40,215 @@ module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
...
+      add_title("GDP control of regional non-CO2 emissions.") %>%
+      add_units("unitless") %>%
+      add_comments("Make year 2010 the control period in creation of GDP control data.") %>%
+      add_comments("Then, created a new column for data that has regional non-CO2 emission species.") %>%
+      add_comments("Then, delete GDP control functions that exist.") %>%
       add_legacy_name("L251.ctrl.delete") %>%
-      add_precursors("precursor1", "precursor2", "etc") %>%
-      # typical flags, but there are others--see `constants.R`
-      add_flags(FLAG_LONG_YEAR_FORM, FLAG_NO_XYEAR) ->
+      add_precursors("temp-data-inject/L161.SSP2_EF", "temp-data-inject/L201.nonghg_steepness", "emissions/A_regions") ->
```
Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: `L201.nonghg_steepness` is also a precursor, right?

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 15 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: YES!  Sorry, I put it back in after rechecking to make sure the L201*** file worked.  It will be removed.

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 16 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: Yes?  It is there, or did I do that wrong?

Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: S. Smith and I are going over the chunk and adding comments about what the code is attempting to do later today.

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 11 above)
Comment (rplzzz) [inline R/zchunk_L251.en_ssp_nonco2.R]: Control-alt-delete.

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 16 above)
Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: It's not in the line above, anyway.

Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: Is it that each precursor has its own line? rather than including all precursors in 1 line?

Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: I mean, there are three precursors listed there, and `L201.nonghg_steepness` isn't one of them.

Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: I am not seeing what you are seeing.  This is the line:
`add_precursors("temp-data-inject/L161.SSP2_EF", "temp-data-inject/L201.nonghg_steepness", "emissions/A_regions") ->`
L201 is the 2nd of the 3 precursors?

Comment (bpbond) [inline R/zchunk_L251.en_ssp_nonco2.R]: I'm having the worst Monday ever. NEVER MIND!

Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: It is alright.  I thought I was crazy.

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 11 above)
Comment (CLynchy) [inline R/zchunk_L251.en_ssp_nonco2.R]: I thought "alt" was a tidyr function I was missing or needed to use.  The joke went over my head.  A joke "is
a witticism, a gag, a bon mot, a fluctuation of words concluding with a trick ending."

File: R/zchunk_L251.en_ssp_nonco2.R (hunk 2 above)
Comment (ssmithClimate) [inline R/zchunk_L251.en_ssp_nonco2.R]: I asked Cary to change those back. 1.5 reads like this is referring to a 1.5 scenario. In some of the text we did change this to be more explicit (e.g. SSP1 and SSP5), but to keep length down some continuation is useful in this particular line.

File: R/zchunk_L251.en_ssp_nonco2.R [hunk 17]
```diff
@@ -1,27 +1,32 @@
...
-#' @export
-module_emissions_L251.en_ssp_nonco2_DISABLED <- function(command, ...) {
+#' @author CDL May 2017
+
+module_emissions_L251.en_ssp_nonco2 <- function(command, ...) {
   if(command == driver.DECLARE_INPUTS) {
-    return(c(FILE = "common/GCAM_region_names",
-             FILE = "emissions/A_regions",
+    return(c(FILE = "emissions/A_regions",
+             FILE = "common/GCAM_region_names",
```
Comment (ssmithClimate) [inline R/zchunk_L251.en_ssp_nonco2.R]: Could you move any emission mapping files from the emissions directory into the emissions/mappings directory?

**Comment by bpbond** [comment]
Looks good, and is ready to go once @ssmithClimate approves I think. One question: "1975" is used a bunch of times throughout code; should this be defined as a constant up top (or potentially in `constants.R`)?

**Comment by CLynchy** [comment]
The 1975 I think is meant to be a placeholder in most cases.  @ssmithClimate ?

**Comment by ssmithClimate** [comment]
Because of a limitation in the core gcam code (since fixed in a pending proposal) all GHG control objects must be read in in 1975 (they are then copied forward by the code). Where it's being used in that context a global constant such as GHG_CONTROL_READIN_YEAR would make sense to me.

**Comment by bpbond** [comment]
@ssmithClimate cool, thanks. @CLynchy could you put a `GHG_CONTROL_READIN_YEAR` into `constants.R` then, and replace "1975" with that throughout?

**Comment by ssmithClimate** [comment]
@bpbond Its fine with me.
